src/SignalR/clients/cpp/src/signalrclient/hub_connection_impl.cpp

```diff
@@ -89,7 +102,49 @@ namespace signalr
 
     pplx::task<void> hub_connection_impl::start()
     {
-        return m_connection->start();
+        if (m_connection->get_connection_state() != connection_state::disconnected)
+        {
+            throw signalr_exception(
+                _XPLATSTR("the connection can only be started if it is in the disconnected state"));
+        }
+
+        m_handshakeTask = pplx::task_completion_event<void>();
+        m_handshakeReceived = false;
+        auto weak_connection = m_connection->weak_from_this();
+        return m_connection->start()
+            .then([weak_connection, this](pplx::task<void> startTask)
+            {
+                startTask.get();
+                auto connection = weak_connection.lock();
+                if (!connection)
+                {
+                    // The connection has been destructed, we should just return and avoid any unobserved task exceptions
+                    return pplx::task_from_result();
```

Comment:
If an invalid handshake response results in a faulted task, so should a connection being deconstructed during start.

Reply:
The difference is that this would only happen if someone loses the reference to the hub_connection, which I believe would crash the process due to an unhandled exception in a task

Reply:
> I believe would crash the process due to an unhandled exception in a task

I don't think I understand. Can you demonstrate the problem?

Callers to start() are going to need to handle faulted task resulting from handshake failures anyway. I would just return a faulted task in this case.

Reply:
Ok fair, even if the user loses the ref to `hub_connection` if they call a method that returns a task they should observe the task
```c++
pplx::task startTask;
// scope
{
    hub_connection connection;
    startTask = connection.start();
}
startTask.get();
```

Reply:
`startTask.get();` could throw even if the hub_connection wasn't destructed due to server issues, network failure, etc... I cannot even begin imagine why you wouldn't want it to throw when hub_connection was destructed.